examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/provisioning/DeviceProvisioningFragment.kt

```diff
@@ -120,7 +126,37 @@ class DeviceProvisioningFragment : Fragment() {
 
       val deviceId = DeviceIdUtil.getNextAvailableId(requireContext())
       val connId = bluetoothManager.connectionId
-      deviceController.pairDevice(gatt, connId, deviceId, deviceInfo.setupPinCode, networkCredentials)
+      deviceController.pairDevice(gatt, connId, deviceId, deviceInfo.setupPinCode, null, networkCredentials, object: DeviceAttestationFailureCallback {
+        override fun onDeviceAttestationFailed(
+          deviceControllerPtr: Long,
+          devicePtr: Long,
+          errorCode: Int
+        ) {
+          requireActivity().runOnUiThread(Runnable {
+            val alertDialog: AlertDialog? = activity?.let {
+              val builder = AlertDialog.Builder(it)
+              builder.apply {
+                setPositiveButton("Continue",
+                  DialogInterface.OnClickListener { dialog, id ->
+                    deviceController.continueCommissioning(devicePtr, true)
+                  })
+                setNegativeButton("No",
+                  DialogInterface.OnClickListener { dialog, id ->
+                    deviceController.continueCommissioning(devicePtr, false)
+                  })
+              }
+              builder.setTitle("Device Attestation")
+              builder.setMessage("Device Attestation failed for device under commissioning. Do you wish to continue pairing?")
```

Comment:
It looks to me the user provided attestation delegate does not conduct the additional attestation with user provided PAAs, it just prompt to the user and ask if the user would like to proceed the pairing if the device under commissioning has been failed to be verified by the default DAC verifier, right?

Reply:
> It looks to me the user provided attestation delegate does not conduct the additional attestation with user provided PAAs, it just prompt to the user and ask if the user would like to proceed the pairing if the device under commissioning has been failed to be verified by the default DAC verifier, right?

Yes, you can see spec **5.5. Commissioning Flows**

> Commissioner SHALL establish the authenticity of the Commissionee as a certified Matter device (see [Section 6.2.3, “Device Attestation Procedure”](https://github.com/project-chip/connectedhomeip/pull/23173#ref_Device_Attestation_Procedure)).

> If the Commissionee fails the Device Attestation Procedure, for any reason, the Commissioner MAY choose to either continue to the Commissioning, or terminate it, depending on implementation-dependent policies.

> Upon failure of the procedure, the Commissioner SHOULD warn the user that the Commissionee is not a fully trusted device, and MAY give the user the choice to authorize or deny the commissioning. Such a warning enables user choice in Commissionee trust on their Fabric, for development workflows, as well as homebrew device development. Such a warning SHOULD contain as much information as the commissioner can provide about the Commissionee, and SHOULD be adapted to the reason of the failure, for example by being different between the case of an expired certificate and the case of a failed signature verification.